lib/spack/llnl/util/filesystem.py

```diff
@@ -1816,66 +1823,67 @@ def find(root, files, recursive=True):
     return result
 
 
-@system_path_filter
-def _find_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict. The idea is that
-    # we want to poke the filesystem as little as possible, but still maintain
-    # stability in the order of the answer. Thus we are recording each library
-    # found in a key, and reconstructing the stable order later.
-    found_files = collections.defaultdict(list)
-
-    # Make the path absolute to have os.walk also return an absolute path
-    root = os.path.abspath(root)
-    for path, _, list_files in os.walk(root):
-        for search_file in search_files:
-            matches = glob.glob(os.path.join(path, search_file))
-            matches = [os.path.join(path, x) for x in matches]
-            found_files[search_file].extend(matches)
+@system_path_filter(arg_slice=slice(1))
+def find_max_depth(root, globs, max_depth=_unset):
+    """Given a set of non-recursive glob file patterns, finds all
+    files matching those patterns up to a maximum specified depth.
 
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+    Does not search in directories below the specified depth.
 
-    return answer
+    If ``globs`` is a list, files matching earlier entries are placed
+    in the return value before files matching later entries.
+    """
+    if not os.path.exists(root):
+        return []
 
+    if isinstance(globs, str):
+        globs = [globs]
 
-@system_path_filter
-def _find_non_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict as os.list_dir
-    # can return files in any order (does not preserve stability)
     found_files = collections.defaultdict(list)
 
-    # Make the path absolute to have absolute path returned
-    root = os.path.abspath(root)
+    dir_queue = collections.deque([(0, root)])
+    while dir_queue:
+        depth, next_dir = dir_queue.pop()
+        if (max_depth is _unset) or depth < max_depth:
+            for dir_entry in os.scandir(next_dir):
+                if dir_entry.is_dir():
+                    dir_path = os.path.join(next_dir, dir_entry.name)
+                    if not os.path.islink(dir_path):
+                        dir_queue.appendleft((depth + 1, os.path.join(next_dir, dir_entry.name)))
 
-    for search_file in search_files:
-        matches = glob.glob(os.path.join(root, search_file))
-        matches = [os.path.join(root, x) for x in matches]
-        found_files[search_file].extend(matches)
+        for glob_pattern in globs:
+            matches = glob.glob(os.path.join(next_dir, glob_pattern))
+            matches = list(os.path.join(next_dir, x) for x in matches)
+            found_files[glob_pattern].extend(matches)
 
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+        # TODO: for fully-recursive searches, we can print a warning after
```

Comment:
Given that (a) we probably need full traversal somewhere internally and (b) we rarely ever need it in packages as library are typically in `./lib`, headers in `./include` etc, wouldn't it make much more sense to expose a `find_*` function in `spack.package.*` that has fixed depth?

Either a wrapper, or just change the `llnl.util.filesystem.*` default to limited depth, and for spack internal use cases where full traversal is necessary disable that.

Reply:
> I feel like limited depth search is the better default in general, don't you think so too?

Yes, I think it's a good default. My concern is what to do when the user explicitly asks for fully recursive search: we don't forbid it, so packages can use it (and so they can still do searches that take "forever", specifically in cases where they are recorded as an external incorrectly).

> The typical package folder structure is <prefix>/{lib,include,share,...}, so depth 1.

You provide an example in https://github.com/spack/spack/pull/41945#discussion_r1447079915 about a `deeply/nested/dir`, so I think the question is how users ought to handle this when there are no symlinks. My guess is that users aren't going to test out different depths: they'll either use our default, or they'll specify a fully-recursive search.

> So, I think if users write find_*(...) in custom libs etc, it might be better to limit that depth, and add a note in the packaging docs.

Are you suggesting removing the option for fully-recursive find from the public interface? My guess is that would be frustrating to users, and we'd have to think hard about what number we limit to as well (possibly drawing inspiration from https://github.com/spack/spack/pull/35902#issuecomment-1463524984).